modules/subschema/src/main/scala/com.snowplowanalytics/iglu.schemaddl/jsonschema/subschema/package.scala

```diff
@@ -3,26 +3,56 @@ package com.snowplowanalytics.iglu.schemaddl.jsonschema.subschema
 import com.snowplowanalytics.iglu.schemaddl.jsonschema.Schema
 import com.snowplowanalytics.iglu.schemaddl.jsonschema.properties.CommonProperties._
 import com.snowplowanalytics.iglu.schemaddl.jsonschema.properties.CommonProperties.Type._
-
 import dregex.Regex
 import io.circe.Json
 
+import scala.annotation.tailrec
+
 package object subschema {
 
   def isSubSchema(s1: Schema, s2: Schema): Compatibility =
     isSubType(simplify(canonicalize(s1)), simplify(canonicalize(s2)))
 
+  @tailrec
   def canonicalize(s: Schema): Schema =
     s match {
-      case s if s.`type`.contains(Null)    => s
-      case s if s.`type`.contains(Boolean) =>
-        if (s.`enum`.isDefined) s else s.copy(`enum` = Some(Enum(List(Json.True, Json.False))))
-      case s if s.`type`.contains(Integer) => s.copy(`type` = Some(Number))
-      case s if s.`type`.contains(Number)  => s
-      case s if s.`type`.contains(String)  => s
+      case s if s.`type`.contains(Null)                => s.copy(`enum` = None)
+      case s if s.`type`.contains(Boolean)             => canonicalizeBoolean(s)
+      case s if s.`type`.contains(Integer)             => s.copy(`type` = Some(Number))
+      case s if s.`type`.contains(Number)              => s
+      case s if s.`type`.contains(String)              => s
+      case s if s.`type`.isEmpty && s.`enum`.isDefined => canonicalize(canonicalizeEnum(s))
       case _ => s
     }
 
+  def canonicalizeBoolean(s: Schema): Schema =
+    if (s.`enum`.isDefined) s else s.copy(`enum` = Some(Enum(List(Json.True, Json.False))))
+
+  def canonicalizeEnum(s: Schema): Schema = {
+    val typeValue: Json => (Type, Option[Json]) =
+      _.fold(
+        jsonNull    = Null -> None,
+        jsonBoolean = v => Boolean -> Some(Json.fromBoolean(v)),
+        jsonNumber  = v => Number -> Some(Json.fromJsonNumber(v)),
+        jsonString  = v => String -> Some(Json.fromString(v)),
+        jsonArray   = v => Array -> Some(Json.fromValues(v)),
+        jsonObject  = v => Object -> Some(Json.fromJsonObject(v))
+      )
+
+    val splitByType: List[Json] => List[Schema] =
+      _.map(typeValue(_))
+        .groupBy(_._1)
+        .mapValues(_.flatMap(_._2))
+        .mapValues(values => if (values.isEmpty) None else Some(Enum(values)))
+        .toList
+        .map({ case (t, e) => Schema.empty.copy(`type` = Some(t), `enum` = e)})
+
+    s.`enum`
+      .map(_.value)
+      .map(splitByType(_))
```

Comment:
enum here is always defined given where we call this from, and also contains at least one element given the schema is valid. 
So we assume it's an etherogeneous array (homogenous array is just a sub-case), and we split into multiple homogenous arrays. 
We finally return a modified schema dropping the old enum array, and adding a disjunction of multiple homogenous enums, making type explicit for all of them.
NOTE: oneOf would be the combinator to use but that would get simplified later so we use anyOf directly given we know all possible intersections of the resulting schemas are empty

Reply:
Given this example of an heterogeneous enum array:
```
{
  "enum": ["red", "amber", "green", null, 42]
}
``` 
When split we will get something like:
```
List(
 List("red", "amber", "green"),
 List(null),
 List(42)
) 
```
And then we apply the `anyOf` because we know there are no intersections between homogeneous lists instead of `oneOf`.

I think it makes sense, just elaborating the reasoning in my head. 😄